# 项目：整理Netflix电影演员评分数据

## 分析目标

此数据分析的目的是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，各演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。

本实战项目的目的在于练习整理数据，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了截止至2022年7月美国地区可观看的所有Netflix电视剧及电影数据。数据集包含两个数据表：`titles.csv`和`credits.csv`。

`titles.csv`包含电影及电视剧相关信息，包括影视作品ID、标题、类型、描述、流派、IMDB（一个国外的在线评分网站）评分，等等。`credits.csv`包含超过7万名出现在Netflix影视作品的导演及演员信息，包括名字、影视作品ID、人物名、演职员类型（导演/演员）等。

`titles.csv`每列的含义如下：
- id：影视作品ID。
- title：影视作品标题。
- show_type：作品类型，电视节目或电影。
- description：简短描述。
- release_year：发布年份。
- age_certification：适龄认证。
- runtime：每集电视剧或电影的长度。
- genres：流派类型列表。
- production_countries：出品国家列表。
- seasons：如果是电视剧，则是季数。
- imdb_id：IMDB的ID。
- imdb_score：IMDB的评分。
- imdb_votes：IMDB的投票数。
- tmdb_popularity：TMDB的流行度。
- tmdb_score：TMDB的评分。

`credits.csv`每列的含义如下：
- person_ID：演职员ID。
- id：参与的影视作品ID。
- name：姓名。
- character_name：角色姓名。
- role：演职员类型，演员或导演。

In [1]:
import pandas as pd

In [2]:
original_titles = pd.read_csv("titles.csv")
original_credits = pd.read_csv("credits.csv")

查看两个文件的前面十条观察值

In [3]:
original_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   
5   ts22164         Monty Python's Flying Circus   SHOW   
6   tm70993                        Life of Brian  MOVIE   
7   tm14873                          Dirty Harry  MOVIE   
8  tm119281                     Bonnie and Clyde  MOVIE   
9   tm98978                      The Blue Lagoon  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   
5  A British sketch comedy series with the shows ...          1969   
6  Brian Cohen is an average young Jewish man, bu...          1979   
7  When a madman dubbed 'Scorpio' terrorizes San ...          1971   
8  In the 1930s, bored waitress Bonnie Parker fal...          1967   
9  Two small children and a ship's cook survive a...          1980   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   
5             TV-14       30                       ['comedy', 'european']   
6                 R       94                                   ['comedy']   
7                 R      102              ['thriller', 'action', 'crime']   
8                 R      110                 ['crime', 'drama', 'action']   
9                 R      104               ['romance', 'action', 'drama']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   
5               ['GB']      4.0  tt0063929         8.8     73424.0   
6               ['GB']      NaN  tt0079470         8.0    395024.0   
7               ['US']      NaN  tt0066999         7.7    155051.0   
8               ['US']      NaN  tt0061418         7.7    112048.0   
9               ['US']      NaN  tt0080453         5.8     69844.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600  
5           17.617       8.306  
6           17.770       7.800  
7           12.817       7.500  
8           15.687       7.500  
9           50.324       6.156

In [5]:
original_credits.head()

person_id       id             name                character   role
0       3748  tm84618   Robert De Niro            Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster            Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                      Tom  ACTOR
3       3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                    Betsy  ACTOR

In [7]:
cleaned_titles = original_titles.copy()
cleaned_credits = original_credits.copy()

开始评估和清洗

In [8]:
cleaned_titles.head()

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600

数据的结构性问题指不符合“每个变量为一列，每个观察值为一行，每种类型的观察单位为一个表格”这三个标准；数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

很明显，titles表中`genres`和`production_countries`列中不满足“每个变量为一列”标准，因此需要进行分割

先查看这两个列的类型

In [9]:
print(type(cleaned_titles['genres'][1]))
print(type(cleaned_titles['production_countries'][1]))

<class 'str'>
<class 'str'>


可知这两个列的数据类型是字符串，应该使用eval函数改成列表再用explode方法进行拆分

In [10]:
print(cleaned_titles['genres'].apply(lambda x:eval(x)))
print(cleaned_titles['production_countries'].apply(lambda x:eval(x)))

0                           [documentation]
1                            [drama, crime]
2       [drama, action, thriller, european]
3                 [fantasy, action, comedy]
4                             [war, action]
                       ...                 
5845                       [romance, drama]
5846                                [drama]
5847                               [comedy]
5848                                     []
5849            [family, animation, comedy]
Name: genres, Length: 5850, dtype: object
0           [US]
1           [US]
2           [US]
3           [GB]
4       [GB, US]
          ...   
5845        [NG]
5846          []
5847        [CO]
5848        [US]
5849          []
Name: production_countries, Length: 5850, dtype: object


重新赋值给`cleaned_titles['genres']`和`['production_countries']`

In [11]:
cleaned_titles['genres'] = cleaned_titles['genres'].apply(lambda x:eval(x))
cleaned_titles['production_countries'] = cleaned_titles['production_countries'].apply(lambda x:eval(x))

检查一下

In [12]:
print(cleaned_titles['genres'][0])
print(cleaned_titles['production_countries'][0])

['documentation']
['US']


都变成列表了，接下来进行拆分

In [13]:
cleaned_titles = cleaned_titles.explode('genres')
cleaned_titles = cleaned_titles.explode('production_countries')
cleaned_titles.head(5)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation                   US      1.0   
1                 R      114          drama                   US      NaN   
1                 R      114          crime                   US      NaN   
2                 R      109          drama                   US      NaN   
2                 R      109         action                   US      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300

至此表cleaned_titles的结构性问题解决，接下来查看cleaned_credits表

In [14]:
cleaned_credits.head()

person_id       id             name                character   role
0       3748  tm84618   Robert De Niro            Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster            Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                      Tom  ACTOR
3       3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                    Betsy  ACTOR

表original_credits不存在结构性问题，双表结构性问题解决。接下来进行内容性分析

### 数据干净度

通过info对数据内容进行大致了解

In [15]:
cleaned_titles.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17818 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    17818 non-null  object 
 1   title                 17817 non-null  object 
 2   type                  17818 non-null  object 
 3   description           17790 non-null  object 
 4   release_year          17818 non-null  int64  
 5   age_certification     10889 non-null  object 
 6   runtime               17818 non-null  int64  
 7   genres                17755 non-null  object 
 8   production_countries  17439 non-null  object 
 9   seasons               6224 non-null   float64
 10  imdb_id               17116 non-null  object 
 11  imdb_score            16976 non-null  float64
 12  imdb_votes            16945 non-null  float64
 13  tmdb_popularity       17663 non-null  float64
 14  tmdb_score            17241 non-null  float64
dtypes: float64(5), int64(2), 

通过info了解到一共有17818条观察值，而除了`id`,`type`,`release_year`,`runtime`四个列不存在空值，其余列均存在空值的情况，将在后续进行评估和清理

此外，我们发现release_year列的数据类型为int类型，年份的数据类型应该为datetime类型，应修改

In [16]:
cleaned_titles['release_year'] = pd.to_datetime(cleaned_titles['release_year'],format = '%Y')
cleaned_titles['release_year']

0      1945-01-01
1      1976-01-01
1      1976-01-01
2      1972-01-01
2      1972-01-01
          ...    
5847   2021-01-01
5848   2021-01-01
5849   2021-01-01
5849   2021-01-01
5849   2021-01-01
Name: release_year, Length: 17818, dtype: datetime64[ns]

接下来对cleaned_credits进行info

In [17]:
cleaned_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  int64 
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.0+ MB


In [18]:
cleaned_credits.head()

person_id       id             name                character   role
0       3748  tm84618   Robert De Niro            Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster            Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                      Tom  ACTOR
3       3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                    Betsy  ACTOR

通过对cleaned_credits进行info后可以发现，character列存在空值，后续对其进行评估和清洗。

此外，person_id为演员id，应为字符串类型，通过astype方法进行修改

In [19]:
cleaned_credits['person_id'] = cleaned_credits['person_id'].astype('str')
cleaned_credits['person_id']

0           3748
1          14658
2           7064
3           3739
4          48933
          ...   
77796     736339
77797     399499
77798     373198
77799     378132
77800    1950416
Name: person_id, Length: 77801, dtype: object

### 数据缺失处理

在`cleaned_titles`中，`title`、`description`、`age_certification`、`genres`、`production_countries`、`seasons`、`imdb_id`、`imdb_score`、`tmdb_popularity`、`tmdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score`变量存在缺失值。

由于影视作品的标题、描述、适龄认证、发行国家、电视剧季数、IMDB的ID、TMDB的流行度、TMDB的评分，并不影响我们挖掘各个流派中的高IMDB评分作品演员，所以可以保留`title`、`description`、`age_certification`、`production_countries`、`seasons`、`imdb_id`、`tmdb_popularity`、`tmdb_score`、`imdb_votes`、`tmdb_popularity`、`tmdb_score`变量值存在空缺的观察值。

因为此数据分析的目的是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，各演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。
所以`imdb_score`和`genres`，即IMDB评分和流派，和我们后续要做的分析息息相关。

In [20]:
cleaned_titles.query('imdb_score.isnull()')

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
75     tm132164                       Bill Hicks: Sane Man  MOVIE   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
...         ...                                        ...    ...   
5810  tm1225897                                 Social Man  MOVIE   
5833   ts307884                                 HQ Barbers   SHOW   
5840  tm1216735                            Sun of the Soil  MOVIE   
5844  tm1074617              Bling Empire - The Afterparty  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   

                                            description release_year  \
0     This collection includes 12 World War II-era p...   1945-01-01   
75    Sane Man was filmed before Bill recorded ‘Dang...   1989-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
...                                                 ...          ...   
5810  Two competitive social media Influencers go he...   2021-01-01   
5833  When a family run barber shop in the heart of ...   2021-01-01   
5840  In 14th-century Mali, an ambitious young royal...   2022-01-01   
5844  The stars of "Bling Empire" discuss the show's...   2021-01-01   
5848  Jamie Foxx, David Alan Grier and more from the...   2021-01-01   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation                   US      1.0   
75                   R       80         comedy                   US      NaN   
145               TV-G       18  documentation                   JP     12.0   
145               TV-G       18         family                   JP     12.0   
145               TV-G       18        reality                   JP     12.0   
...                ...      ...            ...                  ...      ...   
5810               NaN       96          drama                  NaN      NaN   
5833             TV-14       24         comedy                   NG      1.0   
5840               NaN       26            NaN                  NaN      NaN   
5844               NaN       35            NaN                   US      NaN   
5848             PG-13       37            NaN                   US      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
75           NaN         NaN         NaN            3.377         7.5  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
...          ...         ...         ...              ...         ...  
5810  tt20198164         NaN         NaN              NaN         NaN  
5833         NaN         NaN         NaN            0.840         NaN  
5840         NaN         NaN         NaN            1.179         7.0  
5844         NaN         NaN         NaN              NaN         NaN  
5848         NaN         NaN         NaN            1.296        10.0  

[842 rows x 15 columns]

因为以上数据缺失了分析所需的核心数据imdb_score，我们将把这些观察值删除，并查看删除后该列空缺值个数和：

In [21]:
cleaned_titles = cleaned_titles.dropna(subset = ['imdb_score'])
cleaned_titles['imdb_score'].isnull().sum()

0

In [22]:
cleaned_titles.query('genres.isnull()')

id                                              title   type  \
1813    ts77824  My Next Guest Needs No Introduction With David...   SHOW   
1939   ts215037                              Minecraft: Story Mode   SHOW   
2386    ts74805                   A Little Help with Carol Burnett   SHOW   
2658   ts265844                                           #ABtalks   SHOW   
4274  tm1172010                                  The Lockdown Plan  MOVIE   
4648  tm1113921                                           In Vitro  MOVIE   

                                            description release_year  \
1813  TV legend David Letterman teams up with fascin...   2018-01-01   
1939  MInecraft: Story Mode is an interactive, anima...   2018-01-01   
2386  In this unscripted series starring comedy lege...   2018-01-01   
2658  #ABtalks is a YouTube interview show hosted by...   2018-01-01   
4274                                                NaN   2020-01-01   
4648  'In Vitro' is an otherworldly rumination on me...   2019-01-01   

     age_certification  runtime genres production_countries  seasons  \
1813             TV-MA       50    NaN                   US      4.0   
1939             TV-PG       52    NaN                   US      1.0   
2386              TV-G       24    NaN                   US      1.0   
2658             TV-PG       68    NaN                  NaN      1.0   
4274               NaN       49    NaN                  NaN      NaN   
4648               NaN       27    NaN                  NaN      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1813   tt7829834         7.8      5581.0            8.217         7.6  
1939  tt10498322         5.6       347.0              NaN         NaN  
2386   tt7204366         6.3       237.0            1.621         6.2  
2658  tt12635254         9.6         7.0              NaN         NaN  
4274  tt13079112         6.5         NaN              NaN         NaN  
4648  tt10545994         7.7         NaN              NaN         NaN

由于缺失分析所需的核心数据genres，我们将把这些观察值删除，并查看删除后该列空缺值个数和：

In [23]:
cleaned_titles = cleaned_titles.dropna(subset = ['genres'])
cleaned_titles['genres'].isnull().sum()

0

接下来评估cleaned_credits的缺失数据

In [24]:
cleaned_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  object
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: object(5)
memory usage: 3.0+ MB


In [25]:
cleaned_credits.query('character.isnull()')

person_id         id              name character      role
36         3308    tm84618   Martin Scorsese       NaN  DIRECTOR
59        17727   tm154986      John Boorman       NaN  DIRECTOR
106       11475   tm127384       Terry Jones       NaN  DIRECTOR
107       11473   tm127384     Terry Gilliam       NaN  DIRECTOR
162        1063   tm120801    Robert Aldrich       NaN  DIRECTOR
...         ...        ...               ...       ...       ...
77776   2363022  tm1097142  Mohamed El-Arkan       NaN     ACTOR
77777   1827884  tm1097142     Mohamed Bakir       NaN  DIRECTOR
77783    678884  tm1014599      Segun Arinze       NaN     ACTOR
77789   1962840  tm1014599     Seyi Babatope       NaN  DIRECTOR
77800   1950416  tm1059008    Julian Gaviria       NaN  DIRECTOR

[9772 rows x 5 columns]

其中只有character变量存在缺失值。
角色名并不影响我们挖掘各个流派中的高IMDB评分作品演员，并且此变量缺失也有可能因为演职员类别是导演，没有对应角色，因此可以保留character变量值存在空缺的观察值。

#### 处理重复数据

根据数据变量的含义以及内容来看，`cleaned_titles`里不应该存在每个变量值都相同的观察值，因此查看是否存在重复值。

In [26]:
cleaned_titles.duplicated().sum()

0

输出结果为0，说明不存在重复值。

接下来查看`cleaned_credits`数据表是否存在重复值。

In [27]:
cleaned_credits.duplicated().sum()

0

输出结果为0，说明不存在重复值。

#### 处理不一致数据

针对`cleaned_titles`，不一致数据可能存在于`genres`和`production_countries`变量中，我们将查看是否存在多个不同值指代同一流派，以及多个不同值指代同一国家的情况。

In [29]:
cleaned_titles['genres'].value_counts()

genres
drama            3357
comedy           2419
thriller         1446
action           1339
romance          1080
crime            1066
documentation     981
family            769
animation         732
fantasy           727
european          679
scifi             647
horror            438
history           336
music             266
reality           226
war               221
sport             188
western            53
Name: count, dtype: int64

从上面看出，`genres`列里并不存在不一致数据，各个值都在指代不同的流派。但是里面可能还存在空字符串表示的流派，并非有效数据，因此可以进行删除。

删除后，查看`cleaned_titles`里是否还存在`genres`为空字符串的行：

In [31]:
cleaned_titles.query('genres == ""')

Empty DataFrame
Columns: [id, title, type, description, release_year, age_certification, runtime, genres, production_countries, seasons, imdb_id, imdb_score, imdb_votes, tmdb_popularity, tmdb_score]
Index: []

可见不存在空字符串表示的流派。用一样的流程查看production_countries列表

In [32]:
cleaned_titles['production_countries'].value_counts()

production_countries
US    5648
IN    1610
GB    1068
JP    1046
FR     720
      ... 
GT       1
CU       1
LK       1
NP       1
FO       1
Name: count, Length: 108, dtype: int64

由于`value_counts`执行结果中有太多值，Pandas只会默认显示开头和结尾的一些值。要完整展示结果，可以把`display.max_rows`设置为`None`，即取消展示行数上限。

但因为我们只是在当前调用`value_counts`时才需要看完整结果，所以可以结合`option_context`，只更改临时上限。

In [40]:
with pd.option_context('display.max_rows' , None):
    print(cleaned_titles['production_countries'].value_counts())

production_countries
US         5648
IN         1610
GB         1068
JP         1046
FR          720
KR          637
ES          637
CA          608
DE          383
CN          295
MX          264
IT          224
BR          221
AU          217
TR          195
PH          192
AR          150
ID          149
BE          148
TW          133
NG          131
PL          126
ZA          103
NL          102
HK          102
CO           94
EG           93
DK           89
TH           87
SE           81
LB           70
NO           68
AE           52
IE           49
SG           47
XX           43
IL           42
RU           41
CL           35
CH           33
PS           32
BG           31
MY           30
SA           28
AT           28
IS           28
LU           27
NZ           27
PE           26
RO           25
QA           24
CZ           22
JO           19
FI           18
HU           18
UY           15
MA           15
PT           14
KH           10
KW           10
PR            9
PK 

In [46]:
cleaned_titles.loc[cleaned_titles['production_countries'] == 'LB']

id                               title   type  \
24     tm259855       We Are All for the Fatherland  MOVIE   
27     tm200475                   Beirut, Oh Beirut  MOVIE   
32     tm259684                            Whispers  MOVIE   
89      tm52274                     The Little Wars  MOVIE   
89      tm52274                     The Little Wars  MOVIE   
...         ...                                 ...    ...   
5457   ts329711  Love, Life & Everything in Between   SHOW   
5457   ts329711  Love, Life & Everything in Between   SHOW   
5457   ts329711  Love, Life & Everything in Between   SHOW   
5797  tm1046969           Myriam Fares: The Journey  MOVIE   
5797  tm1046969           Myriam Fares: The Journey  MOVIE   

                                            description release_year  \
24    After the 1978 Israeli invasion of Lebanon, ch...   1979-01-01   
27    In the aftermath of the 1967 defeat, four youn...   1975-01-01   
32    At every station, between sites filled with po...   1980-01-01   
89    Beirut resident Soraya is drawn to two men: da...   1982-01-01   
89    Beirut resident Soraya is drawn to two men: da...   1982-01-01   
...                                                 ...          ...   
5457  This black humor pan-Arabic anthology series i...   2022-01-01   
5457  This black humor pan-Arabic anthology series i...   2022-01-01   
5457  This black humor pan-Arabic anthology series i...   2022-01-01   
5797  From pregnancy to album preparations, Lebanese...   2021-01-01   
5797  From pregnancy to album preparations, Lebanese...   2021-01-01   

     age_certification  runtime         genres production_countries  seasons  \
24                 NaN       74  documentation                   LB      NaN   
27                 NaN      110          drama                   LB      NaN   
32                 NaN       93  documentation                   LB      NaN   
89                 NaN      108            war                   LB      NaN   
89                 NaN      108          drama                   LB      NaN   
...                ...      ...            ...                  ...      ...   
5457             TV-14       27         comedy                   LB      1.0   
5457             TV-14       27          drama                   LB      1.0   
5457             TV-14       27        romance                   LB      1.0   
5797               NaN       72  documentation                   LB      NaN   
5797               NaN       72          music                   LB      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
24     tt0170799         6.1        47.0            0.600         5.2  
27     tt0169599         6.4       108.0            1.025         7.2  
32     tt0170803         7.8        78.0            0.600         5.0  
89     tt0084492         6.3       152.0            1.956         5.5  
89     tt0084492         6.3       152.0            1.956         5.5  
...          ...         ...         ...              ...         ...  
5457  tt18253682         5.6        94.0            3.344         6.0  
5457  tt18253682         5.6        94.0            3.344         6.0  
5457  tt18253682         5.6        94.0            3.344         6.0  
5797  tt14773250         2.3       167.0            1.487         NaN  
5797  tt14773250         2.3       167.0            1.487         NaN  

[70 rows x 15 columns]

In [45]:
cleaned_titles.loc[cleaned_titles['production_countries'] == 'Lebanon']

id  title   type  \
546  tm226362  Bosta  MOVIE   

                                           description release_year  \
546  After 15 years in France, Kamal returns to his...   2005-01-01   

    age_certification  runtime genres production_countries  seasons  \
546               NaN      112    war              Lebanon      NaN   

       imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
546  tt0497335         6.4       605.0            1.623         5.0

从以上输出结果来看，出品国家都用两位的国家代码来表示，除了里面存在一个的`Lebanon`值。

`Lebanon`的国家代码是`LB`，出现了70次，说明此处数据不一致。`LB`和`Lebanon`都在表示同一国家，需要进行统一。

把`cleaned_titles`里，`production_countries`的`"LB"`和`"Lebanon"`统一为`LB`，并检查替换后是否还存在`"LB"`：

In [48]:
cleaned_titles['production_countries'] = cleaned_titles['production_countries'].replace({"Lebanon":'LB'})
cleaned_titles.loc[cleaned_titles['production_countries'] == 'Lebanon']

Empty DataFrame
Columns: [id, title, type, description, release_year, age_certification, runtime, genres, production_countries, seasons, imdb_id, imdb_score, imdb_votes, tmdb_popularity, tmdb_score]
Index: []

可见Lebanon以全替换成LB

In [49]:
cleaned_titles['production_countries'].info()

<class 'pandas.core.series.Series'>
Index: 16970 entries, 1 to 5849
Series name: production_countries
Non-Null Count  Dtype 
--------------  ----- 
16670 non-null  object
dtypes: object(1)
memory usage: 265.2+ KB


`production_countries`列中还存在空值，并非有效数据。但由于出品国家并非分析所需的关键信息，所以可以保留出品国家为空的观察值。

接下来查看cleaned_credits中role列，role表示演职员类型，演员或导演。查看是否存在演员或导演之外的数据

In [51]:
cleaned_credits['role'].value_counts()

role
ACTOR       73251
DIRECTOR     4550
Name: count, dtype: int64

从以上输出结果来看，`role`只有两种可能的值，`ACTOR`或`DIRECTOR`，不存在不一致数据。我们可以把这列的类型转换为`Category`，好处是比字符串类型更节约内存空间，也能表明说值的类型有限。

In [53]:
cleaned_credits['role'] = cleaned_credits['role'].astype('category')
cleaned_credits['role']

0           ACTOR
1           ACTOR
2           ACTOR
3           ACTOR
4           ACTOR
           ...   
77796       ACTOR
77797       ACTOR
77798       ACTOR
77799       ACTOR
77800    DIRECTOR
Name: role, Length: 77801, dtype: category
Categories (2, object): ['ACTOR', 'DIRECTOR']

#### 处理无效或错误数据

可以通过DataFrame的`describe`方法，对数值统计信息进行快速了解。

In [55]:
original_titles.describe()

release_year      runtime      seasons   imdb_score    imdb_votes  \
count   5850.000000  5850.000000  2106.000000  5368.000000  5.352000e+03   
mean    2016.417094    76.888889     2.162868     6.510861  2.343938e+04   
std        6.937726    39.002509     2.689041     1.163826  9.582047e+04   
min     1945.000000     0.000000     1.000000     1.500000  5.000000e+00   
25%     2016.000000    44.000000     1.000000     5.800000  5.167500e+02   
50%     2018.000000    83.000000     1.000000     6.600000  2.233500e+03   
75%     2020.000000   104.000000     2.000000     7.300000  9.494000e+03   
max     2022.000000   240.000000    42.000000     9.600000  2.294231e+06   

       tmdb_popularity   tmdb_score  
count      5759.000000  5539.000000  
mean         22.637925     6.829175  
std          81.680263     1.170391  
min           0.009442     0.500000  
25%           2.728500     6.100000  
50%           6.821000     6.900000  
75%          16.590000     7.537500  
max        2274.044000    10.000000

从以上统计信息来看，`original_titles`里不存在脱离现实意义的数值。

`original_credits`由于不包含表示数值含义的变量，因此无需用`describe`检查。

## 整理数据

In [56]:
cleaned_titles

id                               title   type  \
1       tm84618                         Taxi Driver  MOVIE   
1       tm84618                         Taxi Driver  MOVIE   
2      tm154986                         Deliverance  MOVIE   
2      tm154986                         Deliverance  MOVIE   
2      tm154986                         Deliverance  MOVIE   
...         ...                                 ...    ...   
5846   tm898842                         C/O Kaadhal  MOVIE   
5847  tm1059008                             Lokillo  MOVIE   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   

                                            description release_year  \
1     A mentally unstable Vietnam War veteran works ...   1976-01-01   
1     A mentally unstable Vietnam War veteran works ...   1976-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
...                                                 ...          ...   
5846  A heart warming film that explores the concept...   2021-01-01   
5847  A controversial TV host and comedian who has b...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   

     age_certification  runtime     genres production_countries  seasons  \
1                    R      114      drama                   US      NaN   
1                    R      114      crime                   US      NaN   
2                    R      109      drama                   US      NaN   
2                    R      109     action                   US      NaN   
2                    R      109   thriller                   US      NaN   
...                ...      ...        ...                  ...      ...   
5846               NaN      134      drama                  NaN      NaN   
5847               NaN       90     comedy                   CO      NaN   
5849               NaN        7     family                  NaN      1.0   
5849               NaN        7  animation                  NaN      1.0   
5849               NaN        7     comedy                  NaN      1.0   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1      tt0075314         8.2    808582.0           40.965       8.179  
1      tt0075314         8.2    808582.0           40.965       8.179  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
...          ...         ...         ...              ...         ...  
5846  tt11803618         7.7       348.0              NaN         NaN  
5847  tt14585902         3.8        68.0           26.005       6.300  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  

[16970 rows x 15 columns]

In [57]:
cleaned_credits

person_id         id                name                character  \
0          3748    tm84618      Robert De Niro            Travis Bickle   
1         14658    tm84618        Jodie Foster            Iris Steensma   
2          7064    tm84618       Albert Brooks                      Tom   
3          3739    tm84618       Harvey Keitel  Matthew 'Sport' Higgins   
4         48933    tm84618     Cybill Shepherd                    Betsy   
...         ...        ...                 ...                      ...   
77796    736339  tm1059008    Adelaida Buscato                María Paz   
77797    399499  tm1059008  Luz Stella Luengas             Karen Bayona   
77798    373198  tm1059008         Inés Prieto                    Fanny   
77799    378132  tm1059008        Isabel Gaona                   Cacica   
77800   1950416  tm1059008      Julian Gaviria                      NaN   

           role  
0         ACTOR  
1         ACTOR  
2         ACTOR  
3         ACTOR  
4         ACTOR  
...         ...  
77796     ACTOR  
77797     ACTOR  
77798     ACTOR  
77799     ACTOR  
77800  DIRECTOR  

[77801 rows x 5 columns]

对数据的整理，与分析方向紧密相关。此次数据分析目标是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。

那为了能同时获得流派与演员数据，我们需要把`cleaned_titles`和`cleaned_credits`，通过`id`作为键进行连接，因为两个数据表中`id`都是影视作品ID。

In [58]:
titles_with_credits = pd.merge(cleaned_titles,cleaned_credits, on = 'id', how = 'inner')

In [59]:
titles_with_credits.head()

id        title   type  \
0  tm84618  Taxi Driver  MOVIE   
1  tm84618  Taxi Driver  MOVIE   
2  tm84618  Taxi Driver  MOVIE   
3  tm84618  Taxi Driver  MOVIE   
4  tm84618  Taxi Driver  MOVIE   

                                         description release_year  \
0  A mentally unstable Vietnam War veteran works ...   1976-01-01   
1  A mentally unstable Vietnam War veteran works ...   1976-01-01   
2  A mentally unstable Vietnam War veteran works ...   1976-01-01   
3  A mentally unstable Vietnam War veteran works ...   1976-01-01   
4  A mentally unstable Vietnam War veteran works ...   1976-01-01   

  age_certification  runtime genres production_countries  seasons    imdb_id  \
0                 R      114  drama                   US      NaN  tt0075314   
1                 R      114  drama                   US      NaN  tt0075314   
2                 R      114  drama                   US      NaN  tt0075314   
3                 R      114  drama                   US      NaN  tt0075314   
4                 R      114  drama                   US      NaN  tt0075314   

   imdb_score  imdb_votes  tmdb_popularity  tmdb_score person_id  \
0         8.2    808582.0           40.965       8.179      3748   
1         8.2    808582.0           40.965       8.179     14658   
2         8.2    808582.0           40.965       8.179      7064   
3         8.2    808582.0           40.965       8.179      3739   
4         8.2    808582.0           40.965       8.179     48933   

              name                character   role  
0   Robert De Niro            Travis Bickle  ACTOR  
1     Jodie Foster            Iris Steensma  ACTOR  
2    Albert Brooks                      Tom  ACTOR  
3    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR  
4  Cybill Shepherd                    Betsy  ACTOR

由于我们只对挖掘演员的参演作品口碑感兴趣，导演不在我们的分析范围内，因此根据`role`，筛选出类型为`ACTOR`的观察值，供后续分析。

In [62]:
actor_with_titles = titles_with_credits.query('role == "ACTOR"')

为了挖掘出各个流派中的高IMDB评分作品演员，我们需要先根据流派和演员进行分组。

对演员进行分组的时候，选择的是用`person_id`而不是`name`变量，原因是名字容易出现错拼或者重名的情况，演职员ID会比演员姓名更加准确地反映是哪位演员。

In [88]:
groupby_genres_and_person_id = actor_with_titles.groupby(['genres','person_id'])

分组后，我们只需要对`imdb_score`的值进行聚合计算，因此只提取`imdb_score`变量，然后调用`mean`，来计算各个流派影视作品中，每位演员参演作品的平均IMDB评分。

In [66]:
imdb_score_groupby_genres_and_person_id = groupby_genres_and_person_id['imdb_score'].mean()
imdb_score_groupby_genres_and_person_id

genres   person_id
action   1000         6.866667
         100007       7.000000
         100013       6.400000
         100019       6.500000
         100020       6.500000
                        ...   
western  993735       6.500000
         998673       7.300000
         998674       7.300000
         998675       7.300000
         99940        4.000000
Name: imdb_score, Length: 168881, dtype: float64

我们可以调用`reset_index`，对层次化索引进行重置，得到更加规整的DataFrame。

In [67]:
imdb_score_groupby_genres_and_person_id_df = imdb_score_groupby_genres_and_person_id.reset_index()
imdb_score_groupby_genres_and_person_id_df

genres person_id  imdb_score
0        action      1000    6.866667
1        action    100007    7.000000
2        action    100013    6.400000
3        action    100019    6.500000
4        action    100020    6.500000
...         ...       ...         ...
168876  western    993735    6.500000
168877  western    998673    7.300000
168878  western    998674    7.300000
168879  western    998675    7.300000
168880  western     99940    4.000000

[168881 rows x 3 columns]

In [69]:
genres_max_score = imdb_score_groupby_genres_and_person_id_df.groupby('genres')['imdb_score'].max()
genres_max_score

genres
action           9.3
animation        9.3
comedy           9.2
crime            9.5
documentation    9.1
drama            9.5
european         8.9
family           9.3
fantasy          9.3
history          9.1
horror           9.0
music            8.8
reality          8.9
romance          9.2
scifi            9.3
sport            9.1
thriller         9.5
war              8.8
western          8.9
Name: imdb_score, dtype: float64

在我们知道最高分后，可以把以上结果和之前得到的`imdb_score_groupby_genres_and_person_id_df`再次进行连接，得到最高分对应的各个演员ID是什么，也就是这个最高平均分是哪位演员拿到的。

In [70]:
genres_max_score_with_person_id = pd.merge(imdb_score_groupby_genres_and_person_id_df,genres_max_score,on = ['genres','imdb_score'])
genres_max_score_with_person_id

genres person_id  imdb_score
0     action     12790         9.3
1     action      1303         9.3
2     action     21033         9.3
3     action    336830         9.3
4     action     86591         9.3
..       ...       ...         ...
131      war    826547         8.8
132  western     22311         8.9
133  western     28166         8.9
134  western     28180         8.9
135  western     93017         8.9

[136 rows x 3 columns]

这里我们只知道演员的id，要是能知道演员的名字就能更直观。先从cleaned_titles里面提取person_id和name在和genres_max_score_with_person_id进行组合

因为我们需要用演员的id和名字进行连接，所以我们要的名字和id列里面只需要一个名字就行，把重复的删除

In [74]:
cleaned_credits[['person_id','name']].duplicated().value_counts()

False    54589
True     23212
Name: count, dtype: int64

可以发现存在很多名字重复，进行去重

In [77]:
cleaned_credits[['person_id','name']].drop_duplicates().duplicated().value_counts()

False    54589
Name: count, dtype: int64

可见已经没有重复数据，接下来进行赋值连接

In [78]:
actor_id_and_name = cleaned_credits[['person_id','name']].drop_duplicates()
genres_max_score_and_actor_name = pd.merge(genres_max_score_with_person_id,actor_id_and_name, on = 'person_id')
genres_max_score_and_actor_name

genres person_id  imdb_score                name
0       action     12790         9.3         Olivia Hack
1        scifi     12790         9.3         Olivia Hack
2       action      1303         9.3       Jessie Flower
3    animation      1303         9.3       Jessie Flower
4       family      1303         9.3       Jessie Flower
..         ...       ...         ...                 ...
131        war    826547         8.8         Yuto Uemura
132    western     22311         8.9     Koichi Yamadera
133    western     28166         8.9  Megumi Hayashibara
134    western     28180         8.9      Unsho Ishizuka
135    western     93017         8.9            Aoi Tada

[136 rows x 4 columns]

为了把相同流派都排序在一起，我们还可以用`sort_values`方法，把结果里面的行根据`genres`进行排序，然后用`reset_index`把索引重新排序。

索引重新排序后，DataFrame会多出`index`一列，我们可以再把`index`列进行删除。

In [84]:
genres_max_score_and_actor_name = genres_max_score_and_actor_name.sort_values('genres')
genres_max_score_and_actor_name

genres person_id  imdb_score                name
0     action     12790         9.3         Olivia Hack
12    action    336830         9.3     André Sogliuzzo
7     action     21033         9.3          Zach Tyler
17    action     86591         9.3       Cricket Leigh
2     action      1303         9.3       Jessie Flower
..       ...       ...         ...                 ...
131      war    826547         8.8         Yuto Uemura
132  western     22311         8.9     Koichi Yamadera
133  western     28166         8.9  Megumi Hayashibara
134  western     28180         8.9      Unsho Ishizuka
135  western     93017         8.9            Aoi Tada

[136 rows x 4 columns]

排序完后索引乱了，用reset_index()进行索引重置

In [86]:
genres_max_score_and_actor_name = genres_max_score_and_actor_name.reset_index()
genres_max_score_and_actor_name

index   genres person_id  imdb_score                name
0        0   action     12790         9.3         Olivia Hack
1       12   action    336830         9.3     André Sogliuzzo
2        7   action     21033         9.3          Zach Tyler
3       17   action     86591         9.3       Cricket Leigh
4        2   action      1303         9.3       Jessie Flower
..     ...      ...       ...         ...                 ...
131    131      war    826547         8.8         Yuto Uemura
132    132  western     22311         8.9     Koichi Yamadera
133    133  western     28166         8.9  Megumi Hayashibara
134    134  western     28180         8.9      Unsho Ishizuka
135    135  western     93017         8.9            Aoi Tada

[136 rows x 5 columns]

再删除之前乱的索引

In [87]:
genres_max_score_and_actor_name = genres_max_score_and_actor_name.drop('index',axis = 1)
genres_max_score_and_actor_name

genres person_id  imdb_score                name
0     action     12790         9.3         Olivia Hack
1     action    336830         9.3     André Sogliuzzo
2     action     21033         9.3          Zach Tyler
3     action     86591         9.3       Cricket Leigh
4     action      1303         9.3       Jessie Flower
..       ...       ...         ...                 ...
131      war    826547         8.8         Yuto Uemura
132  western     22311         8.9     Koichi Yamadera
133  western     28166         8.9  Megumi Hayashibara
134  western     28180         8.9      Unsho Ishizuka
135  western     93017         8.9            Aoi Tada

[136 rows x 4 columns]

至此，我们获得了一个演员在演绎某流派的最高评分，便于查看该演员在该流派的表现，利于导演甄选拍摄该流派电影的演员

# 保存处理后的数据

In [89]:
genres_max_score_and_actor_name.to_csv('genres_max_score_and_actor_name.csv',index = False)